runtime/lua/vim/filetype/detect.lua

```diff
@@ -0,0 +1,452 @@
+end
+
+function M.mm(path, bufnr)
+
+end
+
+function M.mms(bufnr)
+  for _, line in ipairs(get_lines(bufnr, 1, 20)) do
+    if line:find("^%s*%%") or line:find("^%s*//") or line:find("^%*") then
+      vim.bo[bufnr].filetype = "mmix"
+      return
+    elseif line:find("^%s*#") then
+      vim.bo[bufnr].filetype = "make"
+      return
+    end
+  end
+  vim.bo[bufnr].filetype = "mmix"
+end
+
+function M.mod(path, bufnr)
+
+end
+
+-- This function checks if one of the first five lines start with a dot. In
+-- that case it is probably an nroff file: 'filetype' is set and 1 is returned.
+function M.nroff(bufnr)
+  for _, line in ipairs(get_lines(bufnr, 1, 5)) do
+    if line:find("^%.") then
+      vim.bo[bufnr].filetype = "nroff"
+      return 1
+    end
+  end
+  return 0
+end
+
+function M.perl(path, bufnr)
+
+end
+
+function M.pl(path, bufnr)
+
+end
+
+function M.pp(path, bufnr)
+
+end
+
+function M.prg(path, bufnr)
+
+end
+
+function M.progress_asm(path, bufnr)
+
+end
+
+function M.progress_cweb(bufnr)
+  if vim.g.filetype_w then
+    vim.bo[bufnr].filetype = vim.g.filetype_w
+  else
+    if get_lines(bufnr, 1) == "&ANALYZE" or get_lines(bufnr, 3) == "&GLOBAL-DEFINE" then
+      vim.bo[bufnr].filetype = "progress"
+    else
+      vim.bo[bufnr].filetype = "cweb"
+    end
+  end
+end
+
+function M.progress_pascal(path, bufnr)
+
+end
+
+function M.proto(path, bufnr)
+
+end
+
+function M.r(bufnr)
+  local lines = get_lines(bufnr, 1, 50)
+  -- TODO: \< / \> which match the beginning / end of a word
+  -- Rebol is easy to recognize, check for that first
+  if table.concat(lines):find("[rR][eE][bB][oO][lL]") then
+    vim.bo[bufnr].filetype = "rebol"
+    return
+  end
+
+  for _, line in ipairs(lines) do
+    -- R has # comments
+    if line:find("^%s*#") then
+      vim.bo[bufnr].filetype = "r"
+      return
+    end
+    -- Rexx has /* comments */
+    if line:find("^%s*/%*") then
+      vim.bo[bufnr].filetype = "rexx"
+      return
+    end
+  end
+
+  -- Nothing recognized, use user default or assume R
+  if vim.g.filetype_r then
+    vim.bo[bufnr].filetype = vim.g.filetype_r
+  else
+    -- Rexx used to be the default, but R appears to be much more popular.
+    vim.bo[bufnr].filetype = "r"
+  end
+end
+
+function M.redif(bufnr)
+  for _, line in ipairs(get_lines(bufnr, 1, 5)) do
+    -- TODO: maybe this is too expensive because a new string is created, any thoughts?
+    -- However, it seems much more readable to me than "^[tT][eE]..."
+    if line:lower():find("^template%-type:") then
+      vim.bo[bufnr].filetype = "redif"
+    end
+  end
+end
+
+function M.rules(path, bufnr)
+
+end
+
+-- This function checks the first 25 lines of file extension "sc" to resolve
+-- detection between scala and SuperCollider
+function M.sc(bufnr)
+  -- TODO: it still needs to be discussed if it's ok to use vim.regex in some cases
+  local regex = vim.regex([[[A-Za-z0-9]*\s:\s[A-Za-z0-9]\|var\s<\|classvar\s<\|\^this.*\||\w*|\|+\s\w*\s{\|\*ar\s]])
```

Comment:
Do we need regex for this?

```lua
line:find("[A-Za-z0-9]*%s:%s[A-Za-z0-9]") or line:find("var%s<") or line:find("classvar%s<") or line:find("^this.*") or line:find("|%w*|") or line:find("%+%s%w*%s{") or line:find("%*ar%s")
```

It may be handy to define a helper function for these kinds of patterns. Something like

```lua
local function findany(s, pats)
    for _, v in ipairs(pats) do
        if s:find(v) then
            return true
        end
    end
    return false
end
```

and used like

```lua
findany(line, {"[A-Za-z0-9]*%s:%s[A-Za-z0-9]", "var%s<", "classvar%s<", "^this.*", "|%w*|", "%+%s%w*%s{", "%*ar%s"})
```

Reply:
Sounds like a good idea; that will crop up a few more times (also in `scripts.vim`, which I'd like to add -- later -- as `detect.scripts()`).